src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -411,6 +412,25 @@ bool HeadersMatch(const std::vector<HeaderMatcher>& header_matchers,
   return true;
 }
 
+absl::optional<uint64_t> HeaderHashHelper(
+    const XdsApi::Route::HashPolicy& policy,
+    grpc_metadata_batch* initial_metadata) {
+  GPR_ASSERT(policy.type == XdsApi::Route::HashPolicy::HEADER);
+  std::string concatenated_value;
+  absl::optional<absl::string_view> value = GetMetadataValue(
+      policy.header_name, initial_metadata, &concatenated_value);
+  if (value.has_value()) {
+    if (policy.regex != nullptr) {
+      std::string key(*value);
+      RE2::GlobalReplace(&key, *policy.regex, policy.regex_substitution);
```

Comment:
Maybe extremely unlikely, but this does seem to be vulnerable by applying the RegEx substitution on the concatenated string instead of on individual header values and then concatenating them together. 

IDNK if this worths attention. The code is slightly more complicated if doing the other other way -:)

Reply:
I don't think it makes sense to split it up.  From the xDS perspective, we should be treating the combined string as a single value, because it's possible that a downstream hop has combined the values before we see it anyway.